### DISCUSSION TITLE
MSSQL Operator Fetching Records Clarifications

### DISCUSSION THREAD

**Discussion by emiliocolo** [apache/airflow#21121] [General] [answered]
Thanks for put mssql operator in place, and will be great some clarifications about how to fetch record with the operator itself.

from the examples below so far looks like select can retrieve records from the database, but nothing can be done with the output itself, my understanding is that the operator does not support passing the query output results anywhere, as by design, so the examples are bit confusing, and if i go into some google searching, it points into some xcom solution for some operators, that it seems kind of delta and not supported in all the operators.

I am wrong, and missing something ? I think other SQL operators have similar understanding issues.

Thanks

get_all_description = MsSqlOperator(
    task_id="get_all_description",
    mssql_conn_id='airflow_mssql',
    sql=r"""SELECT description FROM Users;""",
)

get_countries_from_continent = MsSqlOperator(
    task_id="get_countries_from_continent",
    mssql_conn_id='airflow_mssql',
    sql=r"""SELECT * FROM Country where {{ params.column }}='{{ params.value }}';""",
    params={"column": "CONVERT(VARCHAR, continent)", "value": "Asia"},
)
(
    create_table_mssql_task
    >> insert_mssql_hook()
    >> create_table_mssql_from_external_file
    >> populate_user_table
    >> get_all_countries
    >> get_all_description
    >> get_countries_from_continent
)

**Comment by potiuk** [comment, marked as answer]
Running query in operator is mostly for DDL/DML queries not DQL queries. 

If you wish to make a query and do something with the results you should use Hooks rather than operators. In Airlfow Operators are usually "thin" wrapers around Hooks and you are free to run PythonOperator with a callable where you combine the use of multiple Hooks and for example use run Hook to run query and another to write it's results somewhere. 

This is even easier with taskflow API: https://airflow.apache.org/docs/apache-airflow/stable/tutorial_taskflow_api.html where you can do something like (conceptually): 

```
@task
def run_my_task():
    mysql_hook = MySQLHook(params ...)
    result = mysql_hook.run_query("QUERY")
    file= format_result(result)
    s3_hook = S3Hook(parms)
    s3_hook.send(file)
```

Also remember that Airlfow is more for orchestrating external systems rather than running processing internally - you can do it, but this is not where it shines, and usually such procesing is done outside of airflow - airflow just triggers it and waits for result.

**Reply by emiliocolo** [reply]
Great. thank you for the explanation so much, is very helpful

**Reply by txsing** [reply]
The example really make me confused, after wasting lots of time, I finally realize that it is pointless to have a SELECT query in SQL-Operator. It will be better to change this example.

**Reply by potiuk** [reply]
Absolutely @txsing - feel free to change it especially if you can think some better example. Anyone can - change it. Airflow is created by > 2200 contributors - people like you who found something needs improvement - and they did. 

It's just a matter of creating simple PR where the exaple from https://github.com/apache/airflow/blob/main/tests/system/providers/microsoft/mssql/example_mssql.py  . Should be super easy to do and it's an easy way to give back in return for the free software you got. Also super-easy way to become contributor.

Looking forward to your PR !

**Reply by potiuk** [reply]
BTW. You are likely the best person in the world now to correct it - now that you fall into the problem, you probably know best what query would be less misleading.

**Reply by rukaballo** [reply]
> Also remember that Airlfow is more for orchestrating external systems rather than running processing internally - you can do it, but this is not where it shines, and usually such procesing is done outside of airflow - airflow just triggers it and waits for result.

Great explanation, should be on Airflow Documentation! 
One question regarding the quote above, I want to execute a MSSQL Job and that is not the problem, the problem is that I can not check if the Job as runned OK or not, i.e., the Task always ends OK! Is there a parameter to control that or should I build another task with a SQL Sensor to see if everything went OK.  
` run_sp1 = MsSqlOperator(
        task_id='Run_SP1',
        mssql_conn_id='mssql_test_conn_msdb',
        sql=sql_query_exec_sp,
        autocommit=True)`
Thank you

**Reply by potiuk** [reply]
> Great explanation, should be on Airflow Documentation! 

Good idea. Now that you know I am looking forward to your PR adding it - easy way to become contributor (super easy - just click on "Suggest a change on this page" in the docs and PR will be automatically opened for you - you do not need any tools -  GitHub UI (same as writing a comment here) can be used here.

> One question regarding the quote above, I want to execute a MSSQL Job and that is not the problem, the problem is that I can not check if the Job as runned OK or not, i.e., the Task always ends OK! Is there a parameter to control that or should I build another task with a SQL Sensor to see if everything went OK. ` run_sp1 = MsSqlOperator( task_id='Run_SP1', mssql_conn_id='mssql_test_conn_msdb', sql=sql_query_exec_sp, autocommit=True)` Thank you

Unfortunately, Airflow has some 5000 operators in more than 70 providers. Hard to know all of them. Luckily all those operatos have a documentation and even in the documentation there is link to source code that you (or anyone else) can inspect and find out. Usually operator code is dead-simple to understand, it just need an action to do so (and if you ask question about it here more likely anyone who would want to answer it - including yourself - would have to do that - look at the docs, and if in doubt - at the source code. Extending an opertor or even writing your own is first-class citizen so this is always an option.
